File: kotlin/src/main/kotlin/benchmarks/Base64Decode.kt
package benchmarks

import Benchmark
import java.util.Base64

class Base64Decode : Benchmark() {
    private var n: Long = 0
    private lateinit var str2: String
    private lateinit var bytes: ByteArray
    private var resultVal: UInt = 0u

    init {
        n = configVal("size")
    }

    override fun prepare() {
        val str = "a".repeat(n.toInt())
        str2 = Base64.getEncoder().encodeToString(str.toByteArray())
        bytes = Base64.getDecoder().decode(str2)
    }

    override fun run(iterationId: Int) {
        bytes = Base64.getDecoder().decode(str2)
        resultVal += bytes.size.toUInt()  
    }

    override fun checksum(): UInt {
        val str3 = String(bytes.copyOfRange(0, 4))
        val message = "decode ${str2.take(4)}... to ${str3.take(4)}...: $resultVal"
        return Helper.checksum(message)
    }

    override fun name(): String = "Base64Decode"
}

File: scala/src/main/scala/benchmarks/Base64Decode.scala
package benchmarks

import java.util.Base64

class Base64Decode extends Benchmark:
  private var n: Long = 0
  private var str2: String = _
  private var str3: Array[Byte] = _
  private var resultVal: Long = 0L

  override def name(): String = "Base64::Decode"

  override def prepare(): Unit =
    n = configVal("size")
    val str = "a" * n.toInt
    str2 = Base64.getEncoder.encodeToString(str.getBytes)
    str3 = Base64.getDecoder.decode(str2)

  override def run(iterationId: Int): Unit =
    str3 = Base64.getDecoder.decode(str2)
    resultVal += str3.length.toLong

  override def checksum(): Long =
    val _str3 = new String(str3.take(4), "UTF-8")
    val message = s"decode ${str2.take(4)}... to ${_str3.take(4)}...: $resultVal"
    Helper.checksum(message)
